File: ios/VineTrack/App/Reporting/SprayReportRepository.swift
import Foundation
import UIKit
import PostgREST
import Supabase
import CryptoKit

/// Authenticated Spray Report v1 and hourly-weather server paths.
@MainActor
final class SprayReportRepository {
    static let shared: SprayReportRepository = SprayReportRepository()
    private init() {}

    func fetch(tripId: UUID) async throws -> SprayReportPayloadV1 {
        await recoverRows(tripId: tripId)
        struct Request: Encodable { let p_trip_id: UUID }
        return try await SupabaseClientProvider.shared.client
            .rpc("get_spray_report_v1", params: Request(p_trip_id: tripId))
            .execute()
            .value
    }

    /// Resolves each distinct trip independently so multi-record exports never inherit
    /// another trip's weather or row evidence. Failed trips remain available for offline fallback.
    func fetchAll(tripIds: [UUID]) async -> [UUID: SprayReportPayloadV1] {
        var reports: [UUID: SprayReportPayloadV1] = [:]
        var seen: Set<UUID> = []
        for tripId in tripIds where seen.insert(tripId).inserted {
            if let report = try? await fetch(tripId: tripId) { reports[tripId] = report }
        }
        return reports
    }

    func correctMetadata(
        tripId: UUID,
        expectedVersion: Int,
        machineId: UUID?,
        tractorId: UUID?,
        sprayEquipmentId: UUID?,
        operatorUserId: UUID?,
        fuelConsumptionLPerHour: Double?,
        startEngineHours: Double?,
        endEngineHours: Double?
    ) async throws -> SprayReportPayloadV1 {
        struct Request: Encodable {
            let p_operation_id: UUID
            let p_trip_id: UUID
            let p_expected_version: Int
            let p_machine_id: UUID?
            let p_tractor_id: UUID?
            let p_spray_equipment_id: UUID?
            let p_operator_user_id: UUID?
            let p_fuel_consumption_l_per_hour: Double?
            let p_start_engine_hours: Double?
            let p_end_engine_hours: Double?
        }
        struct Response: Decodable { let report: SprayReportPayloadV1 }
        let request = Request(
            p_operation_id: UUID(), p_trip_id: tripId, p_expected_version: expectedVersion,
            p_machine_id: machineId, p_tractor_id: tractorId, p_spray_equipment_id: sprayEquipmentId,
            p_operator_user_id: operatorUserId, p_fuel_consumption_l_per_hour: fuelConsumptionLPerHour,
            p_start_engine_hours: startEngineHours, p_end_engine_hours: endEngineHours
        )
        let response: Response = try await SupabaseClientProvider.shared.client
            .rpc("correct_spray_trip_metadata_v1", params: request)
            .execute()
            .value
        return response.report
    }

    /// Runs the shared server derivation; ambiguous paths are intentionally left unresolved.
    private func recoverRows(tripId: UUID) async {
        guard SupabaseClientProvider.shared.isConfigured,
              let session = try? await SupabaseClientProvider.shared.client.auth.session,
              let url = URL(string: "\(AppConfig.supabaseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/functions/v1/spray-row-recovery") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(AppConfig.supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["tripId": tripId.uuidString.lowercased()])
        _ = try? await URLSession.shared.data(for: request)
    }

    func routeImage(for payload: SprayReportPayloadV1, fallbackTrip: Trip) async -> UIImage? {
        if let route = payload.route,
           route.styleVersion == SprayReportPayloadV1.routeStyleVersion,
           let data = try? await SupabaseClientProvider.shared.client.storage
            .from(route.bucket)
            .download(path: route.objectPath),
           SHA256.hash(data: data).map({ String(format: "%02x", $0) }).joined() == route.sha256,
           let image = UIImage(data: data) {
            return image
        }
        guard let generated = await SprayRecordPDFService.captureMapSnapshot(trip: fallbackTrip) else { return nil }
        return await uploadAndResolveRoute(generated, trip: fallbackTrip) ?? generated
    }

    private func uploadAndResolveRoute(_ image: UIImage, trip: Trip) async -> UIImage? {
        guard let png = image.pngData(), let session = try? await SupabaseClientProvider.shared.client.auth.session else { return nil }
        let routeInput = ([SprayReportPayloadV1.routeStyleVersion, "1030x700"] + trip.pathPoints.map {
            String(format: "%.6f,%.6f", $0.latitude, $0.longitude)
        }).joined(separator: "|")
        let routeHash = SHA256.hash(data: Data(routeInput.utf8)).map { String(format: "%02x", $0) }.joined()
        guard let url = URL(string: "\(AppConfig.supabaseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/functions/v1/spray-report-route-upload") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(AppConfig.supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["tripId": trip.id.uuidString.lowercased(), "routeHash": routeHash, "pngBase64": png.base64EncodedString(), "coordinates": trip.pathPoints.map { ["latitude": $0.latitude, "longitude": $0.longitude] }, "width": 1030, "height": 700])
        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode),
              let envelope = try? JSONDecoder().decode(RouteUploadResponse.self, from: data) else { return nil }
        if envelope.route.sha256 == SHA256.hash(data: png).map({ String(format: "%02x", $0) }).joined() { return image }
        guard let winner = try? await SupabaseClientProvider.shared.client.storage.from(envelope.route.bucket).download(path: envelope.route.objectPath),
              SHA256.hash(data: winner).map({ String(format: "%02x", $0) }).joined() == envelope.route.sha256 else { return nil }
        return UIImage(data: winner)
    }

    private nonisolated struct RouteUploadResponse: Decodable, Sendable { let route: SprayReportPayloadV1.Route }

    /// Creates the scheduled slot even when no genuine provider sample is available.
    /// The server's `(trip_id, sample_slot)` constraint makes retries idempotent.
    func captureUnavailableIfDue(for trip: Trip, at now: Date = Date(), isFinal: Bool = false) async {
        guard trip.tripFunction == TripFunction.spraying.rawValue,
              SupabaseClientProvider.shared.isConfigured else { return }
        let elapsed = max(0, now.timeIntervalSince(trip.startTime))
        let scheduled = isFinal ? now : trip.startTime.addingTimeInterval(floor(elapsed / 3600) * 3600)
        guard let session = try? await SupabaseClientProvider.shared.client.auth.session,
              let url = URL(string: "\(AppConfig.supabaseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/functions/v1/spray-weather-recovery") else { return }
        var edgeRequest = URLRequest(url: url)
        edgeRequest.httpMethod = "POST"
        edgeRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        edgeRequest.setValue(AppConfig.supabaseAnonKey, forHTTPHeaderField: "apikey")
        edgeRequest.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        edgeRequest.httpBody = try? JSONSerialization.data(withJSONObject: ["tripId": trip.id.uuidString.lowercased(), "through": ISO8601DateFormatter().string(from: scheduled)])
        _ = try? await URLSession.shared.data(for: edgeRequest)
        // Failed/transient slots remain absent server-side. Start, resume, restart and
        // final capture all ask the server for the complete missing-slot set again.
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/reporting/SprayReportRepository.kt
package com.rork.vinetrack.data.reporting

import com.rork.vinetrack.data.BackendError
import com.rork.vinetrack.data.SupabaseClient
import com.rork.vinetrack.data.auth.SessionStore
import com.rork.vinetrack.data.model.Trip
import io.ktor.client.call.body
import io.ktor.client.request.headers
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.security.MessageDigest
import java.util.Locale

/** Authenticated Spray Report v1 and hourly-weather server paths. */
class SprayReportRepository(private val session: SessionStore) {
    @Serializable private data class ReportArgs(@SerialName("p_trip_id") val tripId: String)
    @Serializable private data class WeatherArgs(
        @SerialName("p_trip_id") val tripId: String,
        @SerialName("p_sample_slot") val sampleSlot: String,
        @SerialName("p_observed_at") val observedAt: String? = null,
        @SerialName("p_source") val source: String = "Mobile capture unavailable",
        @SerialName("p_source_kind") val sourceKind: String = "unavailable",
        @SerialName("p_station_id") val stationId: String? = null,
        @SerialName("p_temperature_c") val temperatureC: Double? = null,
        @SerialName("p_humidity_pct") val humidityPct: Double? = null,
        @SerialName("p_wind_speed_kmh") val windSpeedKmh: Double? = null,
        @SerialName("p_wind_gust_kmh") val windGustKmh: Double? = null,
        @SerialName("p_wind_direction_deg") val windDirectionDeg: Double? = null,
        @SerialName("p_rain_mm") val rainMm: Double? = null,
        @SerialName("p_is_stale") val isStale: Boolean = false,
    )

    suspend fun fetch(tripId: String): SprayReportPayloadV1 {
        recoverRows(tripId)
        return rpc("get_spray_report_v1", ReportArgs(tripId))
    }

    /** Fetches every distinct trip separately; failures are omitted for caller fallback. */
    suspend fun fetchAll(tripIds: List<String>): Map<String, SprayReportPayloadV1> = buildMap {
        tripIds.distinct().forEach { tripId ->
            runCatching { fetch(tripId) }.getOrNull()?.let { put(tripId, it) }
        }
    }

    @Serializable
    private data class CorrectionArgs(
        @SerialName("p_operation_id") val operationId: String,
        @SerialName("p_trip_id") val tripId: String,
        @SerialName("p_expected_version") val expectedVersion: Long,
        @SerialName("p_machine_id") val machineId: String?,
        @SerialName("p_tractor_id") val tractorId: String?,
        @SerialName("p_spray_equipment_id") val sprayEquipmentId: String?,
        @SerialName("p_operator_user_id") val operatorUserId: String?,
        @SerialName("p_fuel_consumption_l_per_hour") val fuelRate: Double?,
        @SerialName("p_start_engine_hours") val startEngineHours: Double?,
        @SerialName("p_end_engine_hours") val endEngineHours: Double?,
    )
    @Serializable private data class CorrectionResponse(val report: SprayReportPayloadV1)

    suspend fun correctMetadata(
        tripId: String,
        expectedVersion: Long,
        machineId: String?,
        tractorId: String?,
        sprayEquipmentId: String?,
        operatorUserId: String?,
        fuelRate: Double?,
        startEngineHours: Double?,
        endEngineHours: Double?,
    ): SprayReportPayloadV1 {
        val response: CorrectionResponse = rpc(
            "correct_spray_trip_metadata_v1",
            CorrectionArgs(java.util.UUID.randomUUID().toString(), tripId, expectedVersion, machineId, tractorId,
                sprayEquipmentId, operatorUserId, fuelRate, startEngineHours, endEngineHours),
        )
        return response.report
    }

    /** Runs the shared server derivation; ambiguous paths are intentionally left unresolved. */
    private suspend fun recoverRows(tripId: String) {
        if (!SupabaseClient.isConfigured) return
        val token = session.accessToken ?: return
        runCatching {
            SupabaseClient.http.post("${SupabaseClient.baseUrl}/functions/v1/spray-row-recovery") {
                headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
                contentType(ContentType.Application.Json)
                setBody(RowRecoveryArgs(tripId))
            }
        }
    }

    @Serializable private data class RowRecoveryArgs(val tripId: String)

    /** Generates/registers the immutable hybrid route when no canonical winner exists. */
    suspend fun ensureRoute(trip: Trip): SprayReportPayloadV1.Route? {
        val points = trip.pathPoints.orEmpty()
        if (points.size < 2 || !SupabaseClient.isConfigured) return null
        val token = session.accessToken ?: return null
        val input = buildList {
            add(SprayReportPayloadV1.ROUTE_STYLE_VERSION)
            add("1030x700")
            points.forEach { add(String.format(Locale.US, "%.6f,%.6f", it.latitude, it.longitude)) }
        }.joinToString("|")
        val routeHash = MessageDigest.getInstance("SHA-256").digest(input.toByteArray()).joinToString("") { "%02x".format(it) }
        val response = SupabaseClient.http.post("${SupabaseClient.baseUrl}/functions/v1/spray-report-route-upload") {
            headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
            contentType(ContentType.Application.Json)
            setBody(RouteGenerateArgs(trip.id, routeHash, points.map { CoordinateArg(it.latitude, it.longitude) }, 1030, 700))
        }
        if (!response.status.isSuccess()) return null
        return response.body<RouteUploadResponse>().route
    }

    @Serializable private data class CoordinateArg(val latitude: Double, val longitude: Double)
    @Serializable private data class RouteGenerateArgs(val tripId: String, val routeHash: String, val coordinates: List<CoordinateArg>, val width: Int, val height: Int)
    @Serializable private data class RouteUploadResponse(val route: SprayReportPayloadV1.Route)

    suspend fun downloadRoute(route: SprayReportPayloadV1.Route): ByteArray {
        if (!SupabaseClient.isConfigured) throw BackendError.NotConfigured
        val token = session.accessToken ?: throw BackendError.Unauthorized
        val response = SupabaseClient.http.get("${SupabaseClient.baseUrl}/storage/v1/object/authenticated/${route.bucket}/${route.objectPath}") {
            headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
        }
        if (response.status.value == 401 || response.status.value == 403) throw BackendError.Unauthorized
        if (!response.status.isSuccess()) throw BackendError.Server(response.status.value, response.bodyAsText())
        val bytes: ByteArray = response.body()
        val actualHash = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
        if (actualHash != route.sha256) throw BackendError.Server(409, "Canonical route image hash mismatch")
        return bytes
    }

    @Serializable
    data class WeatherRecoveryResult(
        val success: Boolean,
        val captured: Int,
        val unavailable: Int,
        val pending: Int,
        val provider: String? = null,
        val stationId: String? = null,
        val reason: String? = null,
        val errors: List<String> = emptyList(),
    )

    /** Explicit historical recovery. Station availability never controls spray saving. */
    suspend fun recoverWeather(tripId: String, through: Instant): WeatherRecoveryResult {
        if (!SupabaseClient.isConfigured) throw BackendError.NotConfigured
        val token = session.accessToken ?: throw BackendError.Unauthorized
        val response = SupabaseClient.http.post("${SupabaseClient.baseUrl}/functions/v1/spray-weather-recovery") {
            headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
            contentType(ContentType.Application.Json)
            setBody(WeatherRecoveryArgs(tripId, through.toString()))
        }
        if (!response.status.isSuccess()) throw BackendError.Server(response.status.value, response.bodyAsText())
        return response.body()
    }

    /** Recovers every missing scheduled slot; server-side absence is the durable retry queue. */
    suspend fun captureUnavailableIfDue(trip: Trip, now: Instant = Instant.now(), isFinal: Boolean = false) {
        if (trip.tripFunction != "spraying" || !SupabaseClient.isConfigured) return
        val start = trip.startTime?.let { runCatching { Instant.parse(it) }.getOrNull() } ?: return
        val elapsedHours = ChronoUnit.HOURS.between(start, now).coerceAtLeast(0)
        val through = if (isFinal) now else start.plus(elapsedHours, ChronoUnit.HOURS)
        val token = session.accessToken ?: return
        runCatching {
            SupabaseClient.http.post("${SupabaseClient.baseUrl}/functions/v1/spray-weather-recovery") {
                headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
                contentType(ContentType.Application.Json)
                setBody(WeatherRecoveryArgs(trip.id, through.toString()))
            }
        }
    }

    @Serializable private data class WeatherRecoveryArgs(val tripId: String, val through: String)

    private suspend inline fun <reified Body, reified Result> rpc(name: String, body: Body): Result {
        if (!SupabaseClient.isConfigured) throw BackendError.NotConfigured
        val token = session.accessToken ?: throw BackendError.Unauthorized
        val response = SupabaseClient.http.post(SupabaseClient.rpcUrl(name)) {
            headers { append("apikey", SupabaseClient.anonKey); append("Authorization", "Bearer $token") }
            contentType(ContentType.Application.Json)
            setBody(body)
        }
        if (response.status.value == 401 || response.status.value == 403) throw BackendError.Unauthorized
        if (!response.status.isSuccess()) throw BackendError.Server(response.status.value, response.bodyAsText())
        return response.body()
    }
}
